keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlVerificationRepository.kt

```diff
@@ -1,25 +1,120 @@
+  artifactSuppliers,
+  specMigrators
+), VerificationRepository {
+
+  override fun nextEnvironmentsForVerification(
+    minTimeSinceLastCheck: Duration,
+    limit: Int
+  ): Collection<VerificationContext> {
+    val now = clock.instant()
+    val cutoff = now.minus(minTimeSinceLastCheck).toTimestamp()
+    return sqlRetry.withRetry(WRITE) {
+      jooq.inTransaction {
+        select(
+          DELIVERY_CONFIG.UID,
+          DELIVERY_CONFIG.NAME,
+          ENVIRONMENT.UID,
+          ENVIRONMENT.NAME,
+          DELIVERY_ARTIFACT.UID,
+          DELIVERY_ARTIFACT.REFERENCE,
+          ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION
+        )
+          .from(ENVIRONMENT)
+          // join delivery config
+          .join(DELIVERY_CONFIG)
+          .on(DELIVERY_CONFIG.UID.eq(ENVIRONMENT.DELIVERY_CONFIG_UID))
+          // the application is not paused
+          .andNotExists(
+            selectOne()
+              .from(PAUSED)
+              .where(PAUSED.NAME.eq(DELIVERY_CONFIG.APPLICATION))
+              .and(PAUSED.SCOPE.eq(APPLICATION.name))
+          )
+          // join currently deployed artifact version
+          .join(ENVIRONMENT_ARTIFACT_VERSIONS)
+          .on(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(ENVIRONMENT.UID))
+          .and(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS.eq(CURRENT.name))
+          // join artifact
+          .join(DELIVERY_ARTIFACT)
+          .on(DELIVERY_ARTIFACT.UID.eq(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID))
+          // left join so we get results even if there is no row in ENVIRONMENT_LAST_VERIFIED
+          .leftJoin(ENVIRONMENT_LAST_VERIFIED)
+          .on(ENVIRONMENT_LAST_VERIFIED.ENVIRONMENT_UID.eq(ENVIRONMENT.UID))
+          .and(ENVIRONMENT_LAST_VERIFIED.ARTIFACT_UID.eq(DELIVERY_ARTIFACT.UID))
+          // has not been checked recently
+          .where(isnull(ENVIRONMENT_LAST_VERIFIED.AT, LocalDateTime.MIN).lessOrEqual(cutoff))
+          // order by last time checked with things never checked coming first
+          .orderBy(isnull(ENVIRONMENT_LAST_VERIFIED.AT, LocalDateTime.MIN))
+          .limit(limit)
+          .forUpdate()
+          .fetch()
+          .onEach { (_, _, environmentUid, _, artifactUid, _, _) ->
+            insertInto(ENVIRONMENT_LAST_VERIFIED)
+              .set(ENVIRONMENT_LAST_VERIFIED.ENVIRONMENT_UID, environmentUid)
+              .set(ENVIRONMENT_LAST_VERIFIED.ARTIFACT_UID, artifactUid)
+              .set(ENVIRONMENT_LAST_VERIFIED.AT, now.toTimestamp())
+              .onDuplicateKeyUpdate()
+              .set(ENVIRONMENT_LAST_VERIFIED.AT, now.toTimestamp())
+              .execute()
+          }
+          .map { (_, deliveryConfigName, _, environmentName, _, artifactReference, artifactVersion) ->
+            VerificationContext(
+              deliveryConfigByName(deliveryConfigName),
+              environmentName,
+              artifactReference,
+              artifactVersion
+            )
+          }
+      }
+    }
+  }
```

Comment:
This query is not quite as horrible as it looks at first glance. It also manages to avoid the need to pre-seed the "last checked" table whenever we create an environment (see the left join bit). I'd like to migrate that pattern to the similar queries we do elsewhere (separate PR).